src/librustc_metadata/tyencode.rs

```diff
@@ -407,7 +407,14 @@ pub fn enc_existential_bounds<'a,'tcx>(w: &mut Cursor<Vec<u8>>,
 
     enc_region(w, cx, bs.region_bound);
 
-    for tp in &bs.projection_bounds {
+    // Encode projection_bounds in a stable order
+    let mut projection_bounds: Vec<_> = bs.projection_bounds
+                                          .iter()
+                                          .map(|b| (b.item_name().as_str(), b))
+                                          .collect();
+    projection_bounds.sort_by_key(|&(ref name, _)| name.clone());
```

Comment:
Ahh, because `sort_by_key` would make basically O(n log n) calls to its closure, i.e. to `b.item_name().as_str()`? Makes much more sense now (notwithstanding the question of whether it's a premature optimization :). And yeah I overlooked the fact that `sort_by_key` returns unit, so your fix makes sense. Thanks for the explanation! Up to you what changes you want to make, if any.